### PR TITLE
HHH-16853 Setting Named Query Parameters is VERY slow

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-16853

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java [hunk 1]
```diff
@@ -42,19 +43,19 @@ public class ParameterMetadataImpl implements ParameterMetadataImplementor {
...
-		this.names = Collections.emptySet();
-		this.labels = Collections.emptySet();
+		this.queryParametersByNames = Collections.emptyMap();
+		this.queryParametersByLabels = Collections.emptyMap();
 	}
 
 	public ParameterMetadataImpl(Map<QueryParameterImplementor<?>, List<SqmParameter<?>>> queryParameters) {
 		this.queryParameters = queryParameters;
+		this.queryParametersByNames = new HashMap<>();
+		this.queryParametersByLabels = new HashMap<>();
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java:58]: Might be worth only initializing the maps if needed, i.e. if there are named / positional parameters as most of the time neither or only 1 type of parameters is used IMO. Not that big of a deal if it's easier this way.

File: hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java [hunk 2]
```diff
@@ -68,24 +69,20 @@ public ParameterMetadataImpl(Map<QueryParameterImplementor<?>, List<SqmParameter
...
 				labels.add( queryParameter.getPosition() );
+				queryParametersByLabels.put( queryParameter.getPosition(), queryParameter );
 			}
 			else if ( queryParameter.getName() != null ) {
 				if ( names == null ) {
 					names = new HashSet<>();
 				}
-				names.add( queryParameter.getName() );
+				final String name = queryParameter.getName();
+				names.add( name );
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java:79]: Looks like this set and the `labels` set are no longer used and can just be removed.

File: hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java [hunk 3]
```diff
@@ -42,19 +43,19 @@ public class ParameterMetadataImpl implements ParameterMetadataImplementor {
 	public static final ParameterMetadataImpl EMPTY = new ParameterMetadataImpl();
 
 	private final Map<QueryParameterImplementor<?>, List<SqmParameter<?>>> queryParameters;
-
-	private final Set<String> names;
-	private final Set<Integer> labels;
-
+	private final Map<String, QueryParameterImplementor<?>> queryParametersByNames;
+	private final Map<Integer, QueryParameterImplementor<?>> queryParametersByLabels;
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java:47]: I'd prefer something like `queryParametersByName` and `queryParametersByPosition`.

File: hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java [hunk 4]
```diff
@@ -201,7 +195,12 @@ public Set<QueryParameterImplementor<?>> getRegistrations() {
 
 	@Override
 	public boolean hasAnyMatching(Predicate<QueryParameterImplementor<?>> filter) {
-		for ( QueryParameterImplementor<?> queryParameter : queryParameters.keySet() ) {
+		for (QueryParameterImplementor<?> queryParameter : queryParametersByNames.values()  ){
+			if ( filter.test( queryParameter ) ) {
+				return true;
+			}
+		}
+		for (QueryParameterImplementor<?> queryParameter : queryParametersByLabels.values()  ){
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java:203]: Why did you change this? Couldn't we just keep the previous implementation with a single loop?

**Review by mbladel — COMMENTED** [review]
I left some minor comments, other than that LGTM.

File: hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java (hunk 1 above)
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java]: I agree: we can trust that in the majority of use cases, people will be using either named or ordered parameters.

Which implies only one of these `HashMap` instances will be actually used - and since these are expensive to allocate, I'd also prefer to avoid it.

Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java]: P.S. would it be possible to use an array instead of a `Map<Integer, value > ` ?

I think that would be sensible here, as the numbers aren't going to have gaps. The only drawback is that implementation-wise, you'll need to figure out the maximum size of the array before being able to start collecting them; this might imply needing an additional iteration within the constructor, but it's a very fast scan - and you can avoid doing it if there's no positional parameters.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java]: another problem using an array can be the method `Set<Integer> getOrdinalParameterLabels()`,

Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java]: ok, feel free to ignore my remark about Maps and Sets of Integer, we can look into that separately.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java]: anyway I made a change to initialize them only if necessary

**Review by Sanne — APPROVED** [review]
Great, thanks.
Some follow-up ideas, potentially for another day:
 - method `verifyOrdinalParamLabels` could be redesigned; it really doesn't need to allocate two collections and sort them to implement these checks.
 - that `Set<Integer>`, as discussed.
